Fix PolygonSelector.clear()

lschr opened the pull request:
## PR Summary

This PR implements PolygonSelector.clear() to delete all vertices and enable visibility.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

tacaswell commented on lib/matplotlib/widgets.py:
```diff
@@ -3961,6 +3961,12 @@ def verts(self, xys):
+    def clear(self):
+        super().clear()
+        self._xys = [(0, 0)]
+        self.set_visible(True)
```
Why set visible?

tacaswell commented:
Thank you for your work on this and I think overall it makes sense however I have a couple questions.

This looks like this will call `self.update` twice, once through `_draw_polygon` and once through `super().clear()`  Is there a way to get this down to just one call to update (we have a lot of performance issuses that boil down to "we do the same thing N times when once would do")?

Why enable visibilty what the base-class turns it off?

Can you add a test?

lschr commented:
> Why enable visibilty what the base-class turns it off?

Otherwise the selection widget is not visible after calling `clear`. Visibility is also enabled explicitly in `__init__`. I guess `PolygonSelector` is implemented a little differently.

> This looks like this will call `self.update` twice, once through `_draw_polygon` and once through `super().clear()` Is there a way to get this down to just one call to update (we have a lot of performance issuses that boil down to "we do the same thing N times when once would do")?

I guess, it would be smarter to call `self._clear_without_update()` instead of `super().clear()`. However, since this is just

```python
def _clear_without_update(self):
    self._selection_completed = False
    self.set_visible(False)
```

it would also be possible to skip the call and re-enabling visibility altogether, and `PolygonSelector.clear` would become

```python
def clear(self):
    self._selection_completed = False
    self._xys = [(0, 0)]
    self._draw_polygon()
```

What would you suggest?

lschr commented:
I just rebased on current main, updated the code to call `update()` only once and added a test.

@tacaswell can you have another look? I don't think the failing tests are related to this.

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -4141,8 +4141,9 @@ def _on_key_release(self, event):
-    def _draw_polygon(self):
-        """Redraw the polygon based on the new vertex positions."""
+    def _draw_polygon_without_update(self):
+        """Redraw the polygon based on the new vertex positions, no update().
+        """
```
Fits on one line.
```suggestion
        """Redraw the polygon based on the new vertex positions, no update()."""
```

QuLogic commented on lib/matplotlib/tests/test_widgets.py:
```diff
@@ -1686,6 +1686,52 @@ def test_polygon_selector_box(ax):
+    event_sequence = [
+        *polygon_place_vertex(50, 50),
+        *polygon_place_vertex(150, 50),
+        *polygon_place_vertex(50, 150),
+        *polygon_place_vertex(50, 50),
+    ]
+    expected_result = [(50, 50), (150, 50), (50, 150), (50, 50)]
+
+    for (etype, event_args) in event_sequence:
+        do_event(tool, etype, **event_args)
+
+    artist = tool._selection_artist
+
+    assert tool._selection_completed
+    assert tool.get_visible()
+    assert artist.get_visible()
+    np.testing.assert_equal(artist.get_xydata(), expected_result)
+    assert onselect.call_args == ((expected_result[:-1],), {})
+
+    tool.clear()
+    assert not tool._selection_completed
+    np.testing.assert_equal(artist.get_xydata(), [(0, 0)])
+
+    # Do another cycle of events to make sure we can
+    event_sequence = [
+        *polygon_place_vertex(50, 50),
+        *polygon_place_vertex(100, 50),
+        *polygon_place_vertex(50, 150),
+        *polygon_place_vertex(50, 50),
+    ]
+    expected_result2 = [(50, 50), (100, 50), (50, 150), (50, 50)]
+
+    for (etype, event_args) in event_sequence:
+        do_event(tool, etype, **event_args)
+
+    assert tool._selection_completed
+    assert tool.get_visible()
+    assert artist.get_visible()
+    np.testing.assert_equal(artist.get_xydata(), expected_result2)
+    assert onselect.call_args == ((expected_result2[:-1],), {})
```
This could be made a loop over the set of events, something like this:
```python
for result in [[(50, 50), (150, 50), (50, 150), (50, 50)], [...]]:
    for x, y in result:
        for type, event_args in polygon_place_vertex(x, y):
            do_event(tool, etype, **event_args)

    assert stuff with result

    tool.clear()
    assert not tool._selection_completed
    np.testing.assert_equal(artist.get_xydata(), [(0, 0)])
```
